hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -784,29 +788,103 @@ private AbfsOutputStreamContext populateAbfsOutputStreamContext(
             .build();
   }
 
-  public void createDirectory(final Path path, final FsPermission permission,
+  public void createDirectory(final Path path, final FileSystem.Statistics statistics, final FsPermission permission,
       final FsPermission umask, TracingContext tracingContext)
-      throws AzureBlobFileSystemException {
+          throws IOException {
     try (AbfsPerfInfo perfInfo = startTracking("createDirectory", "createPath")) {
-      boolean isNamespaceEnabled = getIsNamespaceEnabled(tracingContext);
-      LOG.debug("createDirectory filesystem: {} path: {} permission: {} umask: {} isNamespaceEnabled: {}",
-              client.getFileSystem(),
-              path,
-              permission,
-              umask,
-              isNamespaceEnabled);
+      if (getAbfsConfiguration().getPrefixMode() == PrefixMode.BLOB) {
+          checkParentChainForFile(path, tracingContext);
 
-      boolean overwrite =
-          !isNamespaceEnabled || abfsConfiguration.isEnabledMkdirOverwrite();
-      final AbfsRestOperation op = client.createPath(getRelativePath(path),
-          false, overwrite,
-              isNamespaceEnabled ? getOctalNotation(permission) : null,
-              isNamespaceEnabled ? getOctalNotation(umask) : null, false, null,
-              tracingContext);
-      perfInfo.registerResult(op.getResult()).registerSuccess(true);
+          HashMap<String, String> metadata = new HashMap<>();
+          metadata.put(X_MS_META_HDI_ISFOLDER, TRUE);
+          createFile(path, statistics, true,
+                 permission, umask, tracingContext, metadata);
+      } else {
+        boolean isNamespaceEnabled = getIsNamespaceEnabled(tracingContext);
+        LOG.debug("createDirectory filesystem: {} path: {} permission: {} umask: {} isNamespaceEnabled: {}",
+                client.getFileSystem(),
+                path,
+                permission,
+                umask,
+                isNamespaceEnabled);
+
+        boolean overwrite =
+                !isNamespaceEnabled || abfsConfiguration.isEnabledMkdirOverwrite();
+        final AbfsRestOperation op = client.createPath(getRelativePath(path),
+                false, overwrite,
+                isNamespaceEnabled ? getOctalNotation(permission) : null,
+                isNamespaceEnabled ? getOctalNotation(umask) : null, false, null,
+                tracingContext);
+        perfInfo.registerResult(op.getResult()).registerSuccess(true);
+      }
     }
   }
 
+  /**
+   * Checks for the entire parent hierarchy and returns if any directory exists and
+   * throws an exception if any file exists.
+   * @param path path to check the hierarchy for.
+   * @param tracingContext the tracingcontext.
+   */
+  private void checkParentChainForFile(Path path, TracingContext tracingContext) throws IOException {
+    if (directoryExists(path, tracingContext)) {
+      return;
+    }
+    for (Path current = path.getParent(), parent = current.getParent();
```

Comment:
let have while loop, reason being so many commands in the for command.

Reply:
Also, it is just looping. What benefit is required from this method.

Reply:
This method will throw an exception if any file is found in the parent hierarchy. The exception logic is written in the checkPathIsDirectory method